### ISSUE TITLE
Difficulty using complex modifications to launch apps

### ISSUE THREAD

**Issue by atjbramley** [pqrs-org/KE-complex_modifications#750]
Hi all, 

Very new to programming and OSX in general so please go easy.  I've got a Logitech K800 which has function keys geared up for Windows use.  I'm trying to use Karabiner to map the function keys to more useful functions.  Some can be done in the function keys window (e.g. map F4 to launchpad) but for a couple I'd like to open apps.  I'd like F2 to open Outlook, F9 to open Music and F3 to open Spotlight.

I tried writing my own complex modification based on the alfred example but can only get a tone out of the button press.

I then downloaded 'Launch apps by right shift+letters' from the library, but these don't work either.  

My thought is that my complex modification may well be right, but that I've not configured OSX to allow Karabiner-Elements to run shell scripts.

Catalina 10.15.6.  Thanks for your help!

**Comment by MuhammedZakir** [comment]
> I then downloaded 'Launch apps by right shift+letters' from the library, but these don't work either.

This is working for me in Mojave. So this is probably due to Catalina incompatibility. Beta version may solve your problem if this is the case -- https://github.com/pqrs-org/Karabiner-Elements/releases/tag/beta.

> My thought is that my complex modification may well be right, but that I've not configured OSX to allow Karabiner-Elements to run shell scripts.
> 
> Catalina 10.15.6. Thanks for your help!x

You can check your config with EventViewer [1]. If your config is correct, it won't show `from.key_code`. For example, if you enable `Launch apps by right shift+letters` rule and you press `right_shift` + `a`, EventViewer will only show `right_shift`.

[1] `EventViewer.app` or `Launch EventViewer` in KarabinerElements menu.

**Comment by atjbramley** [comment]
Some success.  Upgraded to 12.95.1 - the latest beta.

The script as uploaded features 6 different commands - Safari, finder, Terminal, Activity Monitor, Text Edit, Dictionary.

Running the scripts, only Safari and a finder window actually work.  The others don't show the `from.key_code` in event view, but they don't launch the apps either.

I know that the app launches are being triggered by Karabiner because e.g. right-shift + s no longer launches Safari if I disable Karabiner.

I've modifed the scripts a little, without much success.  I've got F3 launching a search in finder (close enough!), but I still want F2 to launch Outlook and F9 to launch music.  These don't work, despite using the same code format as e.g. the code that launches Safari, which does work.

Any ideas where to go now?

**Comment by atjbramley** [comment]
Sorry - new to git too.

**Comment by MuhammedZakir** [comment]
Maybe apps' names/paths are different? I am not sure as I'm using Mojave. See if the path matches with what's shown in `EventViewer > Frontmost application`. You can also check if the command is correct by executing them in terminal - `open '/Applications/Utilities/Activity Monitor.app'`.

**Comment by atjbramley** [comment]
Thanks - the filenames were different and I've got f9 configured for Music now.

I'd still like to point F2 at Outlook but Outlook has spaces in its file path which cause an error in Sublime Text and Karabiner to ignore the complex mod JSON file:

`open /Applications/Microsoft\ Outlook.app`

Again, I've tried a few things, but no success...

**Comment by MuhammedZakir** [comment]
> I'd still like to point F2 at Outlook but Outlook has spaces in its file path which cause an error in Sublime Text and Karabiner to ignore the complex mod JSON file:
> 
> `open /Applications/Microsoft\ Outlook.app`

1.  2 backslashes - `open /Applications/Microsoft\\ Outlook.app`
2. use single quote - `open '/Applications/Microsoft Outlook.app'`

**Comment by atjbramley** [comment]
Not quite Muhammed!  Neither of those cause the json to disappear from the list, which seems to happen when it detects a syntax error, but both of them just produce an error chime.

Would you be able to take one more look?  Here's the full rule, trying the 2 backslash option, in case it helps:

`{
          "type": "basic",
          "from": {
            "key_code": "f2",
            "modifiers": {
              "mandatory": [
                "right_shift"
              ],
              "optional": [
                "caps_lock"
              ]
            }
          },
          "to": [
            {
              "shell_command": "open /Applications/Microsoft\\ Outlook.app"
            }
          ]
        },`

**Comment by MuhammedZakir** [comment]
Yours is indeed correct. I am not sure what's wrong and the error chime is also confusing. I added a rule with wrong path, but there was no chime when I pressed configured key. Config is shown in the list, so no problem with syntax. Other rules work, so probably not a problem with Karabiner. Then, possible problems are application path and, key `f2`.

1. You have probably did it, but check again if path/name matches.
2. If the path is correct, try a different app with `f2`.

If the path is correct and `f2` is working, then I really don't know what is causing this. :-(
